File: 백준/Silver/15654. N과 M （5）/N과 M （5）.py
from sys import stdin
input = stdin.readline

def solution(depth: int):
    if depth == M:
        print(' '.join(map(str, arr)))
        return

    for i in range(N):
        if not visit[i]:
            visit[i] = True
            arr[depth] = nums[i]
            solution(depth + 1)
            visit[i] = False

N, M = map(int, input().split())
nums = list(map(int, input().split()))
nums.sort()
visit = [False] * N
arr = [0] * M

solution(0)

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int N, M;
    static int[] nums;
    static int[] arr;
    static boolean[] visit;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        nums = new int[N];
        for (int i = 0; i < nums.length; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }
        br.close();
        Arrays.sort(nums);

        arr = new int[M];
        visit = new boolean[N];

        solution(0);

        System.out.println(sb);
    }

    static void solution(int depth) {
        if (depth == M) {
            for (int i : arr) {
                sb.append(i).append(' ');
            }
            sb.append('\n');
            return;
        }

        for (int i = 0; i < N; i++) {
            if (!visit[i]) {
                visit[i] = true;
                arr[depth] = nums[i];
                solution(depth + 1);
                visit[i] = false;
            }
        }
    }
}